### PR TITLE
[SMS] service plan id version of the API

### PR DESCRIPTION
**Pull request by 650elx**
To use the other version of the API:

`
sinch_client  = SinchClient(...)
sinch_client.configuration.sms_authentication_method = HTTPAuthentication.SMS_TOKEN.value
`

Ticket for Wiremock migration:
https://tickets.sinch.com/browse/DEVEXP-696

### CODE REVIEW

File: sinch/core/clients/sinch_client_configuration.py
Comment (asein-sinch) [inline sinch/core/clients/sinch_client_configuration.py:44]: You may want to offer the possibility to use other regions than the default one with this protocol too.

**Review by Dovchik — COMMENTED** [review]
looks good for a draft
can you provide in PR description the ways how to configure
* Usual project id SinchClient
* and service_plan_id SinchClient

File: sinch/core/clients/sinch_client_configuration.py
Comment (asein-sinch) [inline sinch/core/clients/sinch_client_configuration.py:86]: I understand that even if the customer is filling the servicePlanId and the API token, he will have to invoke this method on top. Would it be possible to determine which authentication method to use depending on the credentials / region ?

File: sinch/core/enums.py
Comment (asein-sinch) [inline sinch/core/enums.py:16]: Maybe `BEARER_TOKEN` would be more generic as a name for a method that required a bearer token as a value for authentication

Comment (650elx) [reply sinch/core/enums.py:16]: IIRC both `OATUH` and `SMS_TOKEN` are `Bearer` tokens. That's why the naming.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: I think we should discuss this API on our extended discussion meeting.
I'll add this subject to our agenda.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py:48]: I think this comment was already addressed.

Comment (asein-sinch) [reply sinch/core/clients/sinch_client_configuration.py:48]: In this case, I would like to see a test

File: sinch/core/enums.py
Comment (asein-sinch) [reply sinch/core/enums.py:16]: Well, from a code owner point of view, `BEARER_TOKEN` would make more sense as it says what it is. And if I understand correctly, behind the `OAUTH` keyword, there is not a direct `Bearer` with a static token but a logic that check for a JWT existence. So as this one is a standard `Bearer` token usage, it would make sense to refactor the HTTPTransport class to:
 - have a single implementation of `Authorization` header with a `Bearer` token formatting (as you are saying it's used in both OAUTH and SMS_TOKEN)
 - be able to support more APIs with static bearers instead of duplicating the code

**Comment by asein-sinch** [comment]
Is it necessary to target the version 2.0? It will force a user of the SDK already using the SMS with zap stack to do a major update if they want to use a region not supported on zap stack.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: It looks each of us have a different approach to that problem right now.
Until we figure out a unified solution I'm going to keep the code my way.

**Comment by 650elx** [comment]
> Is it necessary to target the version 2.0? It will force a user of the SDK already using the SMS with zap stack to do a major update if they want to use a region not supported on zap stack.

We've already discussed that standup and bumping the major version was chosen as the most convenient and intuitive.

**Comment by JPPortier** [comment]
> We've already discussed that standup and bumping the major version was chosen as the most convenient and intuitive.
I did not remember this conclusion.

An alternative (better?) solution is to provide a smooth transition to SDKs user by NOT removing previous API definitions but marking them as deprecated.
Then both APIS are existing and no need to move to a new major release.
So: a major release will be done in the future when really removing them, but until this real breaking change, SDK's users have time to align with incoming changes

File: sinch/core/clients/sinch_client_configuration.py
Comment (JPPortier) [reply sinch/core/clients/sinch_client_configuration.py]: Yes, this point is still in discussion: we can wait for a status before delivering a new feature.

**Comment by asein-sinch** [comment]
> We've already discussed that standup and bumping the major version was chosen as the most convenient and intuitive.

IIRC, the point discussed during the standup meeting was about the PR #31 (https://github.com/sinch/sinch-sdk-python/pull/31#issuecomment-2302169751) because of the breaking changes introduced in the refactoring of the Verification API. As there are no breaking changes in this PR, it seems more convenient and intuitive for the SDK users to benefit from it without having to upgrade to a major version TBH

**Comment by 650elx** [comment]
@JPPortier @asein-sinch Major version bump was conclusion of me, @krogers0607  and @Dovchik (majority of the Team). We've discussed marking some API methods as deprecated but in the end SemVer solution was chosen. It's also nice to have some incentives for SDK consumers to regularly update their dependencies.

File: sinch/core/enums.py
Comment (650elx) [reply sinch/core/enums.py:16]: `Well, from a code owner point of view, BEARER_TOKEN would make more sense as it says what it is` - From my point of view my naming is much more clearer (subjective thing). Having `OATUH` and `SMS_TOKEN` as separate entities is much more descriptive and in reality `it says what it is`.
When it comes to the code duplication in HTTP transport:
```
        elif endpoint.HTTP_AUTHENTICATION == HTTPAuthentication.SMS_TOKEN.value:
            if not self.sinch.configuration.sms_api_token or not self.sinch.configuration.service_plan_id:
                raise ValidationException(
                    message=(
                        "sms_api_token and service_plan_id are required by this API. "
                        "Those credentials can be obtained from Sinch portal."
                    ),
                    is_from_server=False,
                    response=None
                )
            request_data.headers.update({
                "Authorization": f"Bearer {self.sinch.configuration.sms_api_token}",
                "Content-Type": "application/json"
            })
```
As you can see I use separate Enum values for two versions of Bearer authentication, simply to have code that is cleaner and more descriptive (specialised validation and error message).
When it comes to the code duplication:
```
            request_data.headers.update({
                "Authorization": f"Bearer {self.sinch.configuration.sms_api_token}",
                "Content-Type": "application/json"
            })
```
That's a simple dict `literal` (no logic here). I can move that out to a method or function but in reality the readability will be diminished (jumping across method/function calls). 3 lines of duplicated code is sometimes  a cleaner way.

btw. It looks like you're doing it similar thing, but on much bigger scale in your SDK:
Let's take example of [`packages/voice/src/rest/v1/conferences/conferences-api.ts](https://github.com/sinch/sinch-sdk-node/blob/main/packages/voice/src/rest/v1/conferences/conferences-api.ts#L48)

```
    const headers: { [key: string]: string | undefined } = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
```
This dict/object literal is duplicated 3 times in one file.

**Review by asein-sinch — CHANGES_REQUESTED** [review]
Let's take it offline according to our PR etiquette. This is getting off track

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: I've adapted the code to the draft ADR we have for that 👌

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: test added 👌

**Comment by 650elx** [comment]
> Let's take it offline according to our PR etiquette. This is getting off track

I think we've discussed all of the problems from this PR within our extended discussion meeting. I've fixed all issues/comments.

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py:37]: Oops! I've made a copy/paste typo. It's fixed now.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py:69]: In Python starting a variable name with `_` marks it as sort of private. That was my intention.
Developer is able to access and modify it anyway. It's just a convention.

File: sinch/domains/sms/endpoints/batches/list_batches.py
Comment (650elx) [reply sinch/domains/sms/endpoints/batches/list_batches.py]: It's a synonym :) It works the same. I'll bring old `super` invocation for my convenience :)

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (650elx) [reply tests/e2e/sms/batches/test_batch_sms.py:66]: Wiremock is already doing that, but I can add simple integration/unit test.
e2e tests are e2e tests... :)

Comment (650elx) [reply tests/e2e/sms/batches/test_batch_sms.py:66]: done 👌

File: sinch/domains/sms/endpoints/sms_endpoint.py
Comment (650elx) [reply sinch/domains/sms/endpoints/sms_endpoint.py:19]: By default all SMS API related endpoints have `HTTP_AUTHENTICATION` set to `HTTPAuthentication.OAUTH`. Configuration object is a separate entity.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py:86]: IIRC we have customers that are using custom endpoints for SMS, so the mechanism is legit.

For testing environments:

`sinch_client.configuration.sms_origin = "..."`

File: tests/unit/test_configuration.py
Comment (asein-sinch) [reply tests/unit/test_configuration.py]: More than the copy/paste typo, I think the test itself needs to be revisited: you did an update in the parameters to test without the need to change the assertion, which is generally not a good sign of the quality of the test. Testing the full value rather than a part of it would be preferable.
It would also make more sense to use meaningful test values instead of a long sentence that makes a hostname invalid.

File: sinch/core/clients/sinch_client_configuration.py
Comment (asein-sinch) [reply sinch/core/clients/sinch_client_configuration.py:86]: This is not what I meant: this purpose of this function is to override the default SMS domain, which is exactly that the first line is doing.
But then, calling `self._set_sms_origin()` seems unnecessary as this method will reformat the overridden domain with the region. Do you confirm you expect users to use parameterized  domains as overrides?

File: sinch/core/clients/sinch_client_configuration.py
Comment (asein-sinch) [reply sinch/core/clients/sinch_client_configuration.py]: Then I don't understand the commit https://github.com/sinch/sinch-sdk-python/pull/17/commits/f41bc64829122a003bbd8ec51952eb8d9c01c6d7 which is the exact opposite of the comment where you say your intention is to use the property with an `_`

Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: I've simply forgot to make `sms_origin` private (I've noticed that long after your first comment), hence I've made a decision to make both `sms_origin` and `sms_origin_with_service_plan_id` public.

File: sinch/core/clients/sinch_client_configuration.py
Comment (650elx) [reply sinch/core/clients/sinch_client_configuration.py]: Yes, I'll have to add that to the docs (Docs Team will be notified). Parametrised domains are not necessary.
In case of passing a domain without Python's f string parameter, the format function will not return error, but just return string without region (valid case).

In case of testing environments and customers that want to skip this logic entirely and provide sms origin that was constructed on their own, all that needs to be done is:

`sinch_client.configuration.sms_origin = "..."`
or
`sinch_client.configuration.sms_origin_with_service_plan_id = "..."`

I'll also add that to the docs.

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py]: I've made those tests shorter for the ease of reading.
Although those were great Monty Python quotes...

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (asein-sinch) [inline tests/e2e/sms/batches/test_batch_sms.py:73]: There is no place where we ensure the SendSMSBatchResponse correctly reads the response correctly: it implicitly assumes all the properties are present. I remember requesting numerous times to enhance the tests and we have also decided last September that using wiremock from a badly-named branch on a non-reviewed repo needed to be changed towards the official team's mockserver. I know this PR has been opened a long time before that but there had been multiple changes to it and to quote what I can read in this file: `Now is better than never.` Can you open a PR targeting this branch with e2e tests using the mock server?

**Review by asein-sinch — CHANGES_REQUESTED** [review]

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (650elx) [reply tests/e2e/sms/batches/test_batch_sms.py:73]: Wiremock is validating all the payloads URLs etc. If there's some issues test is being interrupted.
I don't have time for mock server integration; I have to leave it to the next dev.
Btw. This is out of the topic? This PR is about service_plan_id integration.

File: sinch/domains/sms/endpoints/sms_endpoint.py
Comment (650elx) [reply sinch/domains/sms/endpoints/sms_endpoint.py:19]: I simply imported SendBatchSMSEndpoint instead of abstract SMSEndpoint :) No need for HTTP_AUTHENTICATION import anymore.

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py:36]: meh; removed

File: tests/conftest.py
Comment (650elx) [reply tests/conftest.py]: 1. This logic is already present in http_transport
```
        elif endpoint.HTTP_AUTHENTICATION == HTTPAuthentication.SMS_TOKEN.value:
            if not self.sinch.configuration.sms_api_token or not self.sinch.configuration.service_plan_id:
                raise ValidationException(
                    message=(
                        "sms_api_token and service_plan_id are required by this API. "
                        "Those credentials can be obtained from Sinch portal."
                    ),
                    is_from_server=False,
                    response=None
                )
```

2.  Why not; done :)

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py]: Why not :) done

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py]: okey, done

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (asein-sinch) [reply tests/e2e/sms/batches/test_batch_sms.py:73]: I'm not sure to follow: against what wiremock is checking a response payload and do you ensure a property defined in a SDK dataclass is properly deserialized? 

> Btw. This is out of the topic? This PR is about service_plan_id integration.

You are writing new e2e tests in this PR, so instead of investing some time on the deprecated wiremock, it would be preferable to invest it on the new mockserver

File: tests/unit/test_configuration.py
Comment (650elx) [reply tests/unit/test_configuration.py]: This test is already rewritten

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (650elx) [reply tests/e2e/sms/batches/test_batch_sms.py:73]: Let's discuss this on our Extended Discussion meeting today

File: tests/integration/test_http_transport.py
Comment (650elx) [reply tests/integration/test_http_transport.py]: IIRC in some other comment you requested some tests for authorization headers for those two APIs. I've added additional assertions :)

File: tests/integration/test_http_transport.py
Comment (asein-sinch) [inline tests/integration/test_http_transport.py:29]: With this update only this endpoint is covered. If any other endpoint doesn't have the HTTP_AUTHENTICATION defined, it will not work. I would be comfortable with it if the classes were generated but we are not safe from a manual mistake.

File: tests/integration/test_sms_endpoint_credendials_formatting.py
Comment (asein-sinch) [inline tests/integration/test_sms_endpoint_credendials_formatting.py:11]: Why do you need to set this values manually while the SMS endpoint is initialized with the right sinch client?
Same question for the test below.

File: tests/e2e/sms/batches/test_batch_sms.py
Comment (650elx) [reply tests/e2e/sms/batches/test_batch_sms.py:73]: https://tickets.sinch.com/browse/DEVEXP-696

**Pull request merged**
